x-pack/plugins/security/server/routes/usage_collection/authentication_type.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { createHash } from 'crypto';
+
+import { schema } from '@kbn/config-schema';
+
+import type { RouteDefinitionParams } from '..';
+import { wrapIntoCustomErrorResponse } from '../../errors';
+import { createLicensedRouteHandler } from '../licensed_route_handler';
+
+const COUNTER_TYPE = 'SecurityAuthType';
+const MINIMUM_ELAPSED_TIME_HOURS = 12;
+
+export function defineRecordUsageCollectionOnAuthTypeRoutes({
+  getAuthenticationService,
+  router,
+  usageCounter,
+}: RouteDefinitionParams) {
+  router.post(
+    {
+      path: '/internal/security/usage_collection/record_auth_type',
+      validate: {
+        body: schema.nullable(
+          schema.object({
+            auth_type: schema.string(),
+            timestamp: schema.number(),
+            username_hash: schema.string(),
+          })
+        ),
+      },
+    },
+    createLicensedRouteHandler(async (context, request, response) => {
+      try {
+        let timestamp = new Date().getTime();
+        const {
+          auth_type: oldAuthType,
+          username_hash: oldUsernameHash,
+          timestamp: oldTimestamp,
+        } = request.body || {
+          auth_type: '',
+          username_hash: '',
+          timestamp,
+        };
+        const authUser = await getAuthenticationService().getCurrentUser(request);
+        const usernameHash = authUser
+          ? createHash('sha3-256').update(authUser?.username).digest('hex')
+          : '';
+
+        const elapsedTimeInHrs = (timestamp - oldTimestamp) / (1000 * 60 * 60);
+
+        if (
+          !authUser?.authentication_type ||
+          (authUser?.authentication_type && authUser?.authentication_type === '')
+        ) {
+          return response.noContent();
+        }
+
+        if (
+          usageCounter &&
+          (elapsedTimeInHrs >= MINIMUM_ELAPSED_TIME_HOURS ||
+            oldUsernameHash !== usernameHash ||
+            oldAuthType !== authUser?.authentication_type)
+        ) {
+          usageCounter.incrementCounter({
+            counterName: authUser?.authentication_type,
```

Comment:
`authentication_type` won't tell us whether or not the user was authenticated via proxy. For that, I think we need the `authentication_provider` from the user (users who authenticate via the `http` provider are proxy-based)

Additionally, in order to satisfy this example, we need to extract the authentication scheme used (`basic`, `bearer`, `ApiKey`, etc.)

<img width="646" alt="image" src="https://user-images.githubusercontent.com/3493255/145097552-1093d1cf-9808-4ee7-8ec1-a7e2aaa48894.png">

Reply:
> `authentication_type` won't tell us whether or not the user was authenticated via proxy. For that, I think we need the `authentication_provider` from the user (users who authenticate via the `http` provider are proxy-based)

Ah, Xavier and I were discussing this offline --
I think where we were getting tripped up is that we were looking at the docs for [Kibana's configurable authentication providers](https://www.elastic.co/guide/en/kibana/current/kibana-authentication.html), and "http" is not an _authentication provider_ according to the config.

However, in the implementation, "http" _is_ an authentication provider: https://github.com/elastic/kibana/blob/6ecffcc57c9a5b0328257cd25b737140a8471dda/x-pack/plugins/security/server/authentication/providers/http.ts

> Additionally, in order to satisfy this example, we need to extract the authentication scheme used (`basic`, `bearer`, `ApiKey`, etc.)

Ah, so it looks like the scheme is not stored in the `AuthenticatedUser`, so we'll have to grab it from the `request` object by parsing the authorization header: 
https://github.com/elastic/kibana/blob/6ecffcc57c9a5b0328257cd25b737140a8471dda/x-pack/plugins/security/server/authentication/providers/http.ts#L66-L75

I was wondering if we could combine all the info we need into a single string value so we can still use the Core usage collector for this, but it sounds like we might want to be able to analyze these things separately. So maybe that's not a good idea, WDYT?

Reply:
> I was wondering if we could combine all the info we need into a single string value so we can still use the Core usage collector for this, but it sounds like we might want to be able to analyze these things separately. So maybe that's not a good idea, WDYT?

I think it depends™️. If we constrain the schemes to `Basic`, `Bearer`, `ApiKey`, and `Custom`, and _only_ report for the `http` provider, then we could accomplish this using the Core usage collector if we were comfortable using 4 distinct objects (one for each scheme). If that's asking too much of the usage collector, _or_ if we feel that's too limiting for potential future use-cases, then we might want to move forward with a bespoke implementation.

My gut feeling is that the Core usage collector may be too limiting. We wouldn't be able to add any more information to this in the future, and we'd likely need to do something custom for the next authentication data point we want to collect. I'm also cognizant that we may never need this flexibility, but I suspect we will.

Reply:
I also recently heard a question about how to measure Daily Active Users (DAU), which we don't have a good way to do right now.

We currently have 7 auth providers in Kibana (_not_ including the "http" provider). Are we interested in comparing usage of these? 

We could define our own saved object type with separate counters like this:
```js
{
  authLoginUsage: {
    providerTypes: {
      basic: number;
      token: number;
      saml: number;
      oidc: number;
      kerberos: number;
      pki: number;
      anonymous: number;
      total: number;
    },
    realmTypes: {
      reserved: number;
      native: number;
      ldap: number;
      activeDirectory: number;
      pki: number;
      file: number;
      saml: number;
      oidc: number;
      kerberos: number;
      total: number;
    },
    httpAuthSchemes: {
      basic: number;
      bearer: number;
      apiKey: number;
      custom: number;
      total: number;
    }
}
```

However, we would lose some fidelity here because we don't know which providers were used with which realms.

Also, I like how the Core usage collector bins these into separate days so we can measure how these things change over time.

If we don't _really_ need to know what realms are being used, I'm leaning towards suggesting that we define two different Core usage counters:

1. Auth providers types used (excluding HTTP):
   ```js
   {
     "counterName": "basic" | "token" | "saml" | "oidc" | "kerberos" | "pki" | "anonymous",
     "counterType": "SecurityAuthProviderType",
   }
   ```
2. HTTP auth schemes used:
   ```js
   {
     "counterName": "basic" | "bearer" | "apiKey" | "custom",
     "counterType": "SecurityAuthHttpScheme",
   }
   ```

If the user is authenticating using an HTTP header (e.g., sessionless), we'd record both metrics. Otherwise, we'd record only the auth provider type.

This has a few advantages, namely:
* The implementation is almost done, just need a few tweaks
* We get metrics binned on a daily basis for free
* Easy to visualize comparisons like the example above without adding runtime fields

Reply:
> If we don't really need to know what realms are being used, I'm leaning towards suggesting that we define two different Core usage counters

This approach looks reasonable to me!

> If the user is authenticating using an HTTP header (e.g., sessionless), we'd record both metrics. Otherwise, we'd record only the auth provider type

This is true only if we expand `SecurityAuthProviderType` to include the `http` pseudo-provider as a distinct `counterName`. If you authenticate to Kibana using an `Authorization` header, then you will always go through the `http` provider, regardless of what type of credentials you end up providing.

This statement might become less true with the introduction of header-based PKI auth (to support PKI on ESS), but we can cross that bridge when we get there.

Reply:
> This is true only if we expand `SecurityAuthProviderType` to include the `http` pseudo-provider as a distinct `counterName`. If you authenticate to Kibana using an `Authorization` header, then you will always go through the `http` provider, regardless of what type of credentials you end up providing.

Ah, I didn't read the code, I assumed that the `http` provider was backed by the other providers and we'd "translate". Now I see that's not the case, it just tests to see if we can fetch the user with the provided authorization header. 

Yeah sounds good to me, I'll edit my comment accordingly.

> This statement might become less true with the introduction of header-based PKI auth (to support PKI on ESS), but we can cross that bridge when we get there.

:+1:

Reply:
I tried to analyze what combinations of providers, realms and HTTP headers would give us more insight into how Kibana's authentication subsystem is used and ended up with two different lists: minimal and verbose.

It seems we haven't discussed setup when Kibana is used in conjunction with the custom Elasticsearch realms, especially via HTTP authentication. I included counters for these setups too.

I know it's a lot of counters and I still need to check if it's easy to consume this type of data, but I wanted to check with you first if there was reason to not pursue this option initially @jportner @legrego ? 

__Minimal (19 counters, excluded least probable and combined least useful)__:
| Provider (Kibana) | Realm (Elasticsearch)        | Usage Counter      |    
|-------------------|------------------------------|--------------------|    
| Basic             | Native/File/AD/LDAP/Reserved | Basic              |    
| Basic             | Custom                       | Custom             |    
| Token             | Native/File/AD/LDAP/Reserved | Token              |    
| Token             | Custom                       | Token_Custom       |    
| SAML              | SAML                         | SAML               |    
| OIDC              | OIDC                         | OIDC               |    
| Kerberos          | Kerberos                     | Kerberos           |    
| PKI               | PKI                          | PKI                |    
| Anonymous         | Native/File/AD/LDAP/Reserved | Anonymous          |    
| Anonymous         | Anonymous                    | Anonymous_PLEASE_NO|
| Anonymous         | Custom                       | Anonymous_Custom   |    
| Anonymous         | ApiKey                       | Anonymous_ApiKey   |    
| HTTP Basic        | Native/File/AD/LDAP/Reserved | HTTP_Basic         |    
| HTTP Basic        | Custom                       | HTTP_Basic_Custom  |    
| HTTP Bearer       | SAML/OIDC/Kerberos/PKI       | HTTP_Bearer        |    
| HTTP Bearer       | Custom                       | HTTP_Bearer_Custom |    
| HTTP ApiKey       | ApiKey                       | HTTP_ApiKey        |    
| HTTP Custom       | JWT                          | HTTP_Custom_JWT    |    
| HTTP Custom       | *                            | HTTP_Custom        |  

__Verbose (42 counters, almost all possible combinations):__
| Provider (Kibana) | Realm (Elasticsearch) | Usage Counter        |
|-------------------|-----------------------|----------------------|
| Basic             | Native                | Native               |
| Basic             | File                  | File                 |
| Basic             | LDAP                  | LDAP                 |
| Basic             | AD                    | AD                   |
| Basic             | Reserved              | Reserved             |
| Basic             | Custom                | Custom               |
| Token             | Native                | Token_Native         |
| Token             | File                  | Token_File           |
| Token             | LDAP                  | Token_LDAP           |
| Token             | AD                    | Token_AD             |
| Token             | Reserved              | Token_Reserved       |
| Token             | Custom                | Token_Custom         |
| SAML              | SAML                  | SAML                 |
| SAML              | Custom                | SAML_Custom          |
| OIDC              | OIDC                  | OIDC                 |
| OIDC              | Custom                | OIDC_Custom          |
| Kerberos          | Kerberos              | Kerberos             |
| Kerberos          | Custom                | Kerberos_Custom      |
| PKI               | PKI                   | PKI                  |
| PKI               | Custom                | PKI_Custom           |
| Anonymous         | Native                | Anonymous_Native     |
| Anonymous         | File                  | Anonymous_File       |
| Anonymous         | LDAP                  | Anonymous_LDAP       |
| Anonymous         | AD                    | Anonymous_AD         |
| Anonymous         | Reserved              | Anonymous_Reserved   |
| Anonymous         | Anonymous             | Anonymous_PLEASE_NO  |
| Anonymous         | Custom                | Anonymous_Custom     |
| Anonymous         | ApiKey                | Anonymous_ApiKey     |
| HTTP Basic        | Native                | HTTP_Basic_Native    |
| HTTP Basic        | File                  | HTTP_Basic_File      |
| HTTP Basic        | LDAP                  | HTTP_Basic_LDAP      |
| HTTP Basic        | AD                    | HTTP_Basic_AD        |
| HTTP Basic        | Reserved              | HTTP_Basic_Reserved  |
| HTTP Basic        | Custom                | HTTP_Basic_Custom    |
| HTTP Bearer       | SAML                  | HTTP_Bearer_SAML     |
| HTTP Bearer       | OIDC                  | HTTP_Bearer_OIDC     |
| HTTP Bearer       | Kerberos              | HTTP_Bearer_Kerberos |
| HTTP Bearer       | PKI                   | HTTP_Bearer_PKI      |
| HTTP Bearer       | Custom                | HTTP_Bearer_Custom   |
| HTTP ApiKey       | ApiKey                | HTTP_ApiKey          |
| HTTP Custom       | JWT                   | HTTP_Custom_JWT      |
| HTTP Custom       | *                     | HTTP_Custom          |

Reply:
@afharo this PR also won't make 8.2 -- so if we are both targeting 8.3 for this, then would it make more sense to use #121992 for these metrics?